wasm/wasm_source/src/vp_nft.rs

```diff
@@ -0,0 +1,391 @@
+            tag: "v1".to_string(),
+            creator: nft_creator.clone(),
+            vp_code,
+            keys: vec![],
+            opt_keys: vec![],
+            tokens: vec![],
+        });
+
+        tx_env.write_log.commit_tx();
+
+        let vp_env =
+            init_vp_env_from_tx(nft_address.clone(), tx_env, |address| {
+                // Apply transfer in a transaction
+                tx_host_env::nft::mint_tokens(MintNft {
+                    address: nft_address.clone(),
+                    tokens: vec![],
+                    creator: nft_creator.clone(),
+                });
+                tx_host_env::insert_verifier(address)
+            });
+
+        let tx_data: Vec<u8> = vec![];
+        let keys_changed: HashSet<storage::Key> =
+            vp_env.all_touched_storage_keys();
+        let verifiers: HashSet<Address> = vp_env.get_verifiers();
+
+        assert!(validate_tx(tx_data, nft_address, keys_changed, verifiers));
+    }
+
+    /// Test that you can create an nft with tokens
+    #[test]
+    fn test_mint_tokens() {
+        let mut tx_env = TestTxEnv::default();
+        tx::init_tx_env(&mut tx_env);
+
+        let nft_creator = address::testing::established_address_2();
+        let nft_token_owner = address::testing::established_address_1();
+        tx_env.spawn_accounts([&nft_creator, &nft_token_owner]);
+
+        // just a dummy vp, its not used during testing
+        let vp_code =
+            std::fs::read(VP_ALWAYS_TRUE_WASM).expect("cannot load wasm");
+
+        let nft_address = tx_host_env::nft::init_nft(CreateNft {
+            tag: "v1".to_string(),
+            creator: nft_creator.clone(),
+            vp_code,
+            keys: vec![],
+            opt_keys: vec![],
+            tokens: vec![],
+        });
+
+        tx_env.commit_tx_and_block();
+
+        let vp_env = init_vp_env_from_tx(nft_address.clone(), tx_env, |_| {
+            // Apply transfer in a transaction
+            tx_host_env::nft::mint_tokens(MintNft {
+                address: nft_address.clone(),
+                creator: nft_creator.clone(),
+                tokens: vec![NftToken {
+                    id: 1,
+                    values: vec![],
+                    opt_values: vec![],
+                    metadata: "".to_string(),
+                    approvals: vec![],
+                    current_owner: Some(nft_token_owner.clone()),
+                    past_owners: vec![],
+                    burnt: false,
+                }],
+            });
+        });
+
+        let tx_data: Vec<u8> = vec![];
+        let keys_changed: HashSet<storage::Key> =
+            vp_env.all_touched_storage_keys();
+        let verifiers: HashSet<Address> = vp_env.get_verifiers();
+
+        assert!(validate_tx(tx_data, nft_address, keys_changed, verifiers));
+    }
+
+    /// Test that only owner can mint new tokens
+    #[test]
+    fn test_mint_tokens_wrong_owner() {
```

Comment:
there seems to be some non-determinism in the VP with this test. I'm running with `RUST_LOG=debug RUST_BACKTRACE=1 cargo test vp_nft::tests::test_mint_tokens_wrong_owner --features vp_nft -- --nocapture` and occasionally it fails with:
```
thread 'vp_nft::tests::test_mint_tokens_wrong_owner' panicked at 'called `Option::unwrap()` on a `None` value', anoma/vm_env/src/nft.rs:157:41
```

Reply:
ah, didnt notice it! nice catch, thanks :)